[ValueTracking] Handle recursive phis in knownFPClass

davemgreen opened the pull request:
As a follow-on to 113686, this breaks the recursion between phi nodes that have p1 = phi(x, p2) and p2 = phi(y, p1). The knownFPClass can be calculated from the classes of p1 and p2.

Currently includes #113686, see 50686f53fb1069df8ad91dbcb655205a5389a36a for the individual commit.

dtcxzyw commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -5999,6 +5999,21 @@ void computeKnownFPClass(const Value *V, const APInt &DemandedElts,
+        if (auto *IncPhi = dyn_cast<PHINode>(IncValue);
+            IncPhi && IncPhi->getNumIncomingValues() == 2) {
+          if (IncPhi->getIncomingValue(0) == P)
+            IncValue = IncPhi->getIncomingValue(1);
```
Should we update `CtxI` to `IncPhi->getIncomingBlock(1)->getTerminator()`?

goldsteinn commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -5999,6 +5999,21 @@ void computeKnownFPClass(const Value *V, const APInt &DemandedElts,
+            match(IncValue, m_Select(m_Value(), m_Value(V), m_Specific(P))))
+          IncValue = V;
+        // Same around 2-operand phi nodes
+        if (auto *IncPhi = dyn_cast<PHINode>(IncValue);
```
nit: `else if`

goldsteinn commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -5999,6 +5999,21 @@ void computeKnownFPClass(const Value *V, const APInt &DemandedElts,
+            IncPhi && IncPhi->getNumIncomingValues() == 2) {
+          if (IncPhi->getIncomingValue(0) == P)
+            IncValue = IncPhi->getIncomingValue(1);
+          if (IncPhi->getIncomingValue(1) == P)
```
nit: `else if`

davemgreen committed:
[ValueTracking] Handle recursive phis in knownFPClass

As a follow-on to #113686, this breaks the recursion between phi nodes that
have p1 = phi(x, p2) and p2 = phi(y, p1). The knownFPClass can be calculated
from the classes of p1 and p2.
